src/internal.c

```diff
@@ -776,19 +794,231 @@ union wolfSSH_key {
 #endif
 };
 
+
+/*
+ * Identifies the flavor of a key, RSA or ECDSA, and returns the key type ID.
+ * The process is to decode the key as if it was RSA and if that fails try
+ * to load it as if ECDSA. Both public and private keys can be decoded.
+ *
+ * @param in        key to identify
+ * @param inSz      size of key
+ * @param isPrivate indicates private or public key
+ * @param heap      heap to use for memory allocation
+ * @return          keyId as int, WS_MEMORY_E, WS_UNIMPLEMENTED_E
+ */
+static int IdentifyKey(const byte* in, word32 inSz, int isPrivate, void* heap)
+{
+    union wolfSSH_key *key = NULL;
+    int keyId = ID_UNKNOWN;
+    word32 scratch;
+    int ret = WS_SUCCESS;
+    int dynType = isPrivate ? DYNTYPE_PRIVKEY : DYNTYPE_PUBKEY;
+
+    key = (union wolfSSH_key*)WMALLOC(sizeof(union wolfSSH_key), heap, dynType);
+    if (key == NULL) {
+        ret = WS_MEMORY_E;
+    }
+
+#ifndef WOLFSSH_NO_RSA
+    if (ret == WS_SUCCESS) {
+        /* Check RSA key */
+        if (keyId == ID_UNKNOWN) {
+            scratch = 0;
+            ret = wc_InitRsaKey(&key->rsa, NULL);
+
+            if (ret == 0) {
+                if (isPrivate) {
+                    ret = wc_RsaPrivateKeyDecode(in, &scratch,
+                            &key->rsa, inSz);
+                }
+                else {
+                    ret = wc_RsaPublicKeyDecode(in, &scratch,
+                            &key->rsa, inSz);
+                }
+
+                /* If decode was successful, this was an RSA key. */
+                if (ret == 0) {
+                    keyId = ID_SSH_RSA;
+                }
+                /* in case we need to check for another key flavor */
+                ret = WS_SUCCESS;
+            }
+
+            wc_FreeRsaKey(&key->rsa);
+        }
+    }
+#endif
+#ifndef WOLFSSH_NO_ECDSA
+    if (ret == WS_SUCCESS) {
+        if (keyId == ID_UNKNOWN) {
+            scratch = 0;
+            ret = wc_ecc_init_ex(&key->ecc, heap, INVALID_DEVID);
+
+            if (ret == 0) {
+                if (isPrivate) {
+                    ret = wc_EccPrivateKeyDecode(in, &scratch,
+                            &key->ecc, inSz);
+                }
+                else {
+                    ret = wc_EccPublicKeyDecode(in, &scratch,
+                            &key->ecc, inSz);
+                }
+
+                /* If decode was successful, this was an RSA key. */
+                if (ret == 0) {
+                    switch (wc_ecc_get_curve_id(key->ecc.idx)) {
```

Comment:
Might make sense to put this into a function. This snippet is also used in wolfSSH_ReadKey_buffer.

Reply:
Check

Reply:
Looks like this one is still outstanding.

Reply:
IdentifyKey() is the function. wolfSSH_ReadKey_buffer() calls IdentifyKey().

Reply:
I see now. Thanks.